backend/checkmach.ml

```diff
@@ -27,454 +27,813 @@
+    match effect with
+    | V.Bot -> Value.bot
+    | V.Safe -> dst
+    | V.Top -> Value.transform dst
+
+  let transform_diverge ~(effect : V.t) (dst : Value.t) =
+    let div = V.join effect dst.div in
+    { dst with div }
+
+  let transform t ~next ~exn ~(effect : Value.t) desc dbg =
+    let next = transform_return ~effect:effect.nor next in
+    let exn = transform_return ~effect:effect.exn exn in
+    report t next ~msg:"transform new next" ~desc dbg;
+    report t exn ~msg:"transform new exn" ~desc dbg;
+    let r = Value.join next exn in
+    report t r ~msg:"transform join" ~desc dbg;
+    let r = transform_diverge ~effect:effect.div r in
+    report t r ~msg:"transform_call result" ~desc dbg;
+    check t r desc dbg
+
+  let transform_call t ~next ~exn callee ~desc dbg =
+    report t next ~msg:"transform_call next" ~desc dbg;
+    report t exn ~msg:"transform_call exn" ~desc dbg;
+    let callee_value, new_dep = find_callee t callee in
+    update_deps t callee_value new_dep desc dbg;
+    transform t ~next ~exn ~effect:callee_value desc dbg
+
+  let transform_operation t (op : Mach.operation) ~next ~exn dbg =
     match op with
     | Imove | Ispill | Ireload | Iconst_int _ | Iconst_float _ | Iconst_symbol _
     | Iload _ | Icompf _ | Inegf | Iabsf | Iaddf | Isubf | Imulf | Idivf
-    | Ifloatofint | Iintoffloat | Ivalueofint | Iintofvalue
+    | Ifloatofint | Iintoffloat
     | Iintop_imm
         ( ( Iadd | Isub | Imul | Imulh _ | Idiv | Imod | Iand | Ior | Ixor
           | Ilsl | Ilsr | Iasr | Ipopcnt | Iclz _ | Ictz _ | Icomp _ ),
           _ )
     | Iintop
         ( Iadd | Isub | Imul | Imulh _ | Idiv | Imod | Iand | Ior | Ixor | Ilsl
         | Ilsr | Iasr | Ipopcnt | Iclz _ | Ictz _ | Icomp _ )
-    | Icsel _ | Iname_for_debugger _ ->
-      assert (Mach.operation_is_pure op)
+    | Ivalueofint | Iintofvalue | Icsel _ | Iname_for_debugger _ ->
+      assert (Mach.operation_is_pure op);
+      assert (not (Mach.operation_can_raise op));
+      next
     | Istackoffset _ | Iprobe_is_enabled _ | Iopaque | Ibeginregion | Iendregion
     | Iintop_atomic _ ->
-      ()
-    | Istore _ -> ()
+      assert (not (Mach.operation_can_raise op));
+      next
+    | Istore _ ->
+      assert (not (Mach.operation_can_raise op));
+      next
     | Iintop Icheckbound | Iintop_imm (Icheckbound, _) ->
-      report_fail t "checkbound" dbg
-    | Ialloc { mode = Alloc_local; _ } -> ()
-    | Ialloc { mode = Alloc_heap; _ } -> report_fail t "allocation" dbg
-    | Ipoll _ ->
-      (* Polling points may trigger finalisers, signal handlers or memprof
-         callbacks, but any allocations done therein don't count towards the
-         calling function being "allocating". *)
-      ()
+      (* does not allocate even when it raises because checkbound exception is
+         static. *)
+      transform t ~next ~exn ~effect:Value.safe "checkbound" dbg
+    | Ipoll _ (* CR gyorsh: poll points are considered allocations. *)
```

Comment:
This pattern should either be moved
to the `Alloc_heap` case, or have code
that ensure we will revisit it when poll
points are generated (_i.e._ `assert false`
or `Misc.fatal`).

Reply:
Fixed.